packages/rest-crud/src/crud-rest-builder.plugin.ts

```diff
@@ -0,0 +1,112 @@
+} from '@loopback/repository';
+import {Model} from '@loopback/rest';
+import debugFactory from 'debug';
+import {defineCrudRepositoryClass, defineCrudRestController} from '.';
+
+const debug = debugFactory('loopback:boot:crud-rest');
+
+export interface ModelCrudRestApiConfig extends ModelApiConfig {
+  // E.g. '/products'
+  basePath: string;
+}
+
+@bind(asModelApiBuilder)
+export class CrudRestApiBuilder implements ModelApiBuilder {
+  readonly pattern: string = 'CrudRest';
+
+  build(
+    application: ApplicationWithRepositories,
+    modelClass: typeof Model & {prototype: Model},
+    cfg: ModelApiConfig,
+  ): Promise<void> {
+    const modelName = modelClass.name;
+    const config = cfg as ModelCrudRestApiConfig;
+    if (!config.basePath) {
+      throw new Error(
+        `Missing required field "basePath" in configuration for model ${modelName}.`,
+      );
+    }
+
+    if (!(modelClass.prototype instanceof Entity)) {
+      throw new Error(
+        `CrudRestController requires an Entity, Models are not supported. (Model name: ${modelName})`,
+      );
+    }
+    const entityClass = modelClass as typeof Entity & {prototype: Entity};
+
+    // TODO Check if the repository class has been already defined.
+    // If yes, then skip creation of the default repository
+    const repositoryClass = setupCrudRepository(entityClass, config);
+    application.repository(repositoryClass);
+    debug('Registered repository class', repositoryClass.name);
+
+    const controllerClass = setupCrudRestController(entityClass, config);
+    application.controller(controllerClass);
+    debug('Registered controller class', controllerClass.name);
+
+    return Promise.resolve();
+  }
+}
+
+/**
+ * Set up a CRUD Repository class for the given Entity class.
+ *
+ * @param entityClass - the Entity class the repository is built for
+ * @param modelConfig - configuration of the Entity class
+ */
+function setupCrudRepository(
+  entityClass: typeof Entity & {prototype: Entity},
+  modelConfig: ModelCrudRestApiConfig,
+): Class<EntityCrudRepository<Entity, unknown>> {
+  const repositoryClass = defineCrudRepositoryClass(entityClass);
+
+  inject(`datasources.${modelConfig.dataSource}`)(
+    repositoryClass,
+    undefined,
+    0,
+  );
+
+  return repositoryClass;
+}
+
+/**
+ * Set up a CRUD Controller class for the given Entity class.
+ *
+ * @param entityClass - the Entity class the controller is built for
+ * @param modelConfig - configuration of the Entity class
+ */
+function setupCrudRestController(
+  entityClass: typeof Entity & {prototype: Entity},
+  modelConfig: ModelCrudRestApiConfig,
+): ControllerClass {
+  const controllerClass = defineCrudRestController(
+    entityClass,
+    // important - forward the entire config object to allow controller
+    // factories to accept additional (custom) config options
+    modelConfig,
+  );
+
+  inject(`repositories.${entityClass.name}Repository`)(
+    controllerClass,
+    undefined,
+    0,
```

Comment:
@nabdelgadir  Can you explain how the `0` is used? It is not clear in the function that receives this value. thx.

Reply:
Somehow I am not able to comment in https://github.com/strongloop/loopback-next/pull/4589#discussion_r377742892, so I am continuing the discussion here.

> Can we create a constant and use that instead of 0? Perhaps also one for the undefined value? So both constants can provide some clarity to the method call?

```
inject(`datasources.${modelConfig.dataSource}`)(
    repositoryClass,
    UNDEFINED_CLASS_CONSTRUCTOR_FUNCTION,
    INDEX_OF_INJECTED_DATASOURCE_PARAMETER_INSIDE_REPOSITORY_CONSTRUCTOR,
```

Personally, I prefer to create a named function for setting up injection of the first constructor argument.

```ts
// usage
injectFirstConstructorArg(controllerClass, `repositories.${entityClass.name}Repository`);
injectFirstConstructorArg(repositoryClass, `datasources.${config.dataSource}`);

// implementation
function injectFirstConstructorArg(ctor, key) {
  inject(key)(ctor, undefined /* constructor member */, 0 /* the first argument */);
}
```

That way the code inside `setupCrudRepository` and `setupCrudRestController` can stay high-level and still communicate the intent well. People interested in implementation details of `injectFirstConstructorArgument` can look into function body, at which point they should be prepared to encounter code assuming knowledge of low-level decorator details.